source/common/router/header_parser.cc

```diff
@@ -342,25 +355,63 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
     headers.remove(header);
   }
 
-  for (const auto& [key, entry] : headers_to_add_) {
-    const std::string value =
-        stream_info != nullptr
-            ? entry.formatter_->format(request_headers, response_headers, *stream_info)
-            : entry.original_value_;
-    if (!value.empty() || entry.add_if_empty_) {
-      switch (entry.append_action_) {
-        PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
-      case HeaderValueOption::APPEND_IF_EXISTS_OR_ADD:
-        headers.addReferenceKey(key, value);
-        break;
-      case HeaderValueOption::ADD_IF_ABSENT:
-        if (auto header_entry = headers.get(key); header_entry.empty()) {
-          headers.addReferenceKey(key, value);
+  if (stream_info != nullptr) {
+    // Create local copy of headers to add and replace. This is required
+    // to execute all formatters using the original received headers.
+    // Only after all the formatters produced the new values of the headers, the headers are set.
+    std::vector<std::pair<const Http::LowerCaseString&, const std::string>> headers_to_add,
+        headers_to_overwrite;
+    for (const auto& [key, entry] : headers_to_add_) {
+      const std::string value =
+          entry.formatter_->format(request_headers, response_headers, *stream_info);
+      if (!value.empty() || entry.add_if_empty_) {
+        switch (entry.append_action_) {
+          PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
+        case HeaderValueOption::APPEND_IF_EXISTS_OR_ADD:
+          headers_to_add.emplace_back(key, value);
+          break;
+        case HeaderValueOption::ADD_IF_ABSENT:
+          if (auto header_entry = headers.get(key); header_entry.empty()) {
+            headers_to_add.emplace_back(key, value);
+          }
+          break;
+        case HeaderValueOption::OVERWRITE_IF_EXISTS_OR_ADD:
+          headers_to_overwrite.emplace_back(key, value);
+          break;
+        }
+      }
+    }
+
+    // First overwrite all headers which need to be overwritten.
+    for (const auto& header : headers_to_overwrite) {
+      headers.setReferenceKey(header.first, header.second);
+    }
+
+    // Now add headers which should be added.
+    for (const auto& header : headers_to_add) {
+      headers.addReferenceKey(header.first, header.second);
+    }
+  } else {
+    // The logic below is very similar to code executed when stream_info != nullptr,
+    // It is possible to combine two branches of "if" statement into more elegant code,
+    // but based on performance testing, using separate branches depending on value
+    // of stream_info resulted in the most efficient code.
```

Comment:
oh I didn't realize we still had a complex loop to run if stream_info was null; I was thinking this meant  we could short-circuit entirely.

Reply:
WDYT of factoring out lines 367-382, and lines 400-416, which are basically the same, into a parameterized helper method or lambda that takes the value as a formal.

Reply:
I have actually already done it (lambdas). The code looks more elegant, but the performance is worse. I am not sure if this is related to capturing params or by introducing extra function calls:

  ```
std::vector<std::pair<const Http::LowerCaseString&, const std::string>> headers_to_add;
  std::vector<std::pair<const Http::LowerCaseString&, const std::string>> headers_to_overwrite;

  std::function<void(const Http::LowerCaseString&, const HeadersToAddEntry&)> header_append_func,
      header_overwrite_func;
  if (stream_info == nullptr) {
    header_append_func = [&headers](const Http::LowerCaseString& key,
                                    const HeadersToAddEntry& entry) {
      if (!entry.original_value_.empty() || entry.add_if_empty_) {
        headers.addReferenceKey(key, entry.original_value_);
      }
    };
    header_overwrite_func = [&headers](const Http::LowerCaseString& key,
                                       const HeadersToAddEntry& entry) {
      if (!entry.original_value_.empty() || entry.add_if_empty_) {
        headers.setReferenceKey(key, entry.original_value_);
      }
    };
  } else {
    header_append_func = [&request_headers, &response_headers, &stream_info, &headers_to_add](
                             const Http::LowerCaseString& key, const HeadersToAddEntry& entry) {
      const std::string value =
          entry.formatter_->format(request_headers, response_headers, *stream_info);
      if (!value.empty() || entry.add_if_empty_) {
        headers_to_add.emplace_back(key, value);
      }
    };
    header_overwrite_func = [&request_headers, &response_headers, &stream_info,
                             &headers_to_overwrite](const Http::LowerCaseString& key,
                                                    const HeadersToAddEntry& entry) {
      const std::string value =
          entry.formatter_->format(request_headers, response_headers, *stream_info);
      if (!value.empty() || entry.add_if_empty_) {
        headers_to_overwrite.emplace_back(key, value);
      }
    };
  }

  for (const auto& [key, entry] : headers_to_add_) {
    switch (entry.append_action_) {
      PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
    case HeaderValueOption::APPEND_IF_EXISTS_OR_ADD:
      header_append_func(key, entry);
      break;
    case HeaderValueOption::ADD_IF_ABSENT:
      if (auto header_entry = headers.get(key); header_entry.empty()) {
        header_append_func(key, entry);
      }
      break;
    case HeaderValueOption::OVERWRITE_IF_EXISTS_OR_ADD:
      header_overwrite_func(key, entry);
      break;
    }
  }

  // First overwrite all headers which need to be overwritten.
  for (const auto& header : headers_to_overwrite) {
    headers.setReferenceKey(header.first, header.second);
  }

  // Now add headers which should be added.
  for (const auto& header : headers_to_add) {
    headers.addReferenceKey(header.first, header.second);
  }
```

So, for valid stream_info the results are:
when there are 2 separate paths with quite similar code `if (stream_info != null) {switch: ...} else {switch: ...}`
```
bmEvaluateHeaders/2        8767 ns         8759 ns        78696
bmEvaluateHeaders/4       21864 ns        21835 ns        31680
bmEvaluateHeaders/6       34422 ns        34383 ns        20185
bmEvaluateHeaders/8       47098 ns        47038 ns        14951
bmEvaluateHeaders/10      59249 ns        59183 ns        11892
bmEvaluateHeaders/12      72207 ns        72126 ns         9793
bmEvaluateHeaders/14      83693 ns        83585 ns         8450
bmEvaluateHeaders/16      95415 ns        95343 ns         7336
bmEvaluateHeaders/18     107372 ns       107341 ns         6471
bmEvaluateHeaders/20     120339 ns       120320 ns         5736
```
and when using lambdas:
```
bmEvaluateHeaders/2        9905 ns         9904 ns        70521
bmEvaluateHeaders/4       23442 ns        23440 ns        29858
bmEvaluateHeaders/6       35770 ns        35756 ns        19602
bmEvaluateHeaders/8       48154 ns        48152 ns        14576
bmEvaluateHeaders/10      60328 ns        60320 ns        11551
bmEvaluateHeaders/12      72932 ns        72918 ns         9601
bmEvaluateHeaders/14      84671 ns        84649 ns         8287
bmEvaluateHeaders/16      96296 ns        96254 ns         7235
bmEvaluateHeaders/18     109546 ns       109532 ns         6387
bmEvaluateHeaders/20     122509 ns       122487 ns         5653
```

And when stream_info is nullptr:
when there are 2 separate paths with quite similar code `if (stream_info != null) {switch: ...} else {switch: ...}`
```
bmEvaluateHeaders/2        6699 ns         6694 ns       104734
bmEvaluateHeaders/4       17388 ns        17387 ns        40493
bmEvaluateHeaders/6       27539 ns        27535 ns        25916
bmEvaluateHeaders/8       38423 ns        38417 ns        18217
bmEvaluateHeaders/10      48765 ns        48759 ns        14305
bmEvaluateHeaders/12      60458 ns        60448 ns        11471
bmEvaluateHeaders/14      69731 ns        69722 ns        10092
bmEvaluateHeaders/16      79537 ns        79528 ns         8744
bmEvaluateHeaders/18      91209 ns        91205 ns         7470
bmEvaluateHeaders/20     102027 ns       102022 ns         6784
```
and with lambdas:
```
bmEvaluateHeaders/2        7157 ns         7154 ns        97919
bmEvaluateHeaders/4       18311 ns        18306 ns        38525
bmEvaluateHeaders/6       28535 ns        28529 ns        24947
bmEvaluateHeaders/8       39269 ns        39262 ns        17732
bmEvaluateHeaders/10      50440 ns        50416 ns        13879
bmEvaluateHeaders/12      60673 ns        60662 ns        11518
bmEvaluateHeaders/14      70968 ns        70952 ns         9895
bmEvaluateHeaders/16      81705 ns        81686 ns         8635
bmEvaluateHeaders/18      92027 ns        92003 ns         7501
bmEvaluateHeaders/20     103374 ns       103325 ns         6796
```
So, the current code with `if (stream_info != nullptr) {switch: ....} else {switch: ....}` gives the most performance. Switching to lambdas adds around (5-10)% for small number of headers and about 2% for 20 headers.

I like lambdas as the code is more elegant. I also measured timing using code compiled with `dbg` flag, so maybe optimized code results in less of performance loss.

Reply:
Thanks -- Can you run the benchmarks with just "-c opt"? Sorry for the trouble :)

I think the use of the lambdas will cause more stack-frames to be generated, and you'd want to see those while debugging, but it's possible the optimizer would remove them if it doesn't have to keep around all the stack-frames for debugging. E.g. typically in flame graphs on optimized code you won't see the entire call-stack.

Reply:
I re-run all tests using optimized code and timings are completely different. Thanks for suggesting it. The PR introduces a small delay (because it needs to temporarily store results in a vector while new headers values are calculated), but the delay is not significant. I also replaced `std::vector` in favour of `absl::InlinedVector` and it improved the timings. lambdas are still a bit slower than using if (stream_info) {...} else {...}, but the difference is small now. The exact numbers are below. Thanks for analyzing my changes with such detail. This code is executed for each HTTP packet, so I am totally OK with trying different approached and profiling it.

```
OPTIMIZED BUILD
===============

stream_info is not nullptr

original:
bmEvaluateHeaders/2         239 ns          239 ns      2933249
bmEvaluateHeaders/4        1058 ns         1058 ns       660594
bmEvaluateHeaders/6        1758 ns         1758 ns       498900
bmEvaluateHeaders/8        2093 ns         2093 ns       422586
bmEvaluateHeaders/10       2275 ns         2274 ns       326849
bmEvaluateHeaders/12       2621 ns         2621 ns       271865
bmEvaluateHeaders/14       3042 ns         3041 ns       227186
bmEvaluateHeaders/16       3538 ns         3538 ns       196645
bmEvaluateHeaders/18       3968 ns         3968 ns       174172
bmEvaluateHeaders/20       4558 ns         4558 ns       157163

this PR (if (strem_info == nullptr) {...} else {...}
bmEvaluateHeaders/2         255 ns          255 ns      2754158
bmEvaluateHeaders/4         992 ns          992 ns       703809
bmEvaluateHeaders/6        1688 ns         1688 ns       442525
bmEvaluateHeaders/8        2053 ns         2053 ns       374797
bmEvaluateHeaders/10       2544 ns         2543 ns       288706
bmEvaluateHeaders/12       2985 ns         2985 ns       235778
bmEvaluateHeaders/14       3489 ns         3489 ns       202675
bmEvaluateHeaders/16       3984 ns         3983 ns       178978
bmEvaluateHeaders/18       4514 ns         4514 ns       156524
bmEvaluateHeaders/20       5096 ns         5094 ns       134769

this PR lambdas
bmEvaluateHeaders/2         324 ns          324 ns      2152446
bmEvaluateHeaders/4        1031 ns         1031 ns       664590
bmEvaluateHeaders/6        1729 ns         1729 ns       430220
bmEvaluateHeaders/8        2131 ns         2131 ns       361016
bmEvaluateHeaders/10       2623 ns         2623 ns       281483
bmEvaluateHeaders/12       3046 ns         3045 ns       226996
bmEvaluateHeaders/14       3598 ns         3597 ns       197573
bmEvaluateHeaders/16       4044 ns         4044 ns       173756
bmEvaluateHeaders/18       4697 ns         4696 ns       153256
bmEvaluateHeaders/20       5319 ns         5318 ns       129045

stream_info is nullptr

original (before PR):
bmEvaluateHeaders/2         164 ns          164 ns      4256833
bmEvaluateHeaders/4         710 ns          710 ns       979517
bmEvaluateHeaders/6        1183 ns         1183 ns       695658
bmEvaluateHeaders/8        1539 ns         1539 ns       629900
bmEvaluateHeaders/10       1663 ns         1663 ns       483293
bmEvaluateHeaders/12       1973 ns         1973 ns       403045
bmEvaluateHeaders/14       2147 ns         2147 ns       338610
bmEvaluateHeaders/16       2617 ns         2617 ns       295247
bmEvaluateHeaders/18       2702 ns         2702 ns       252027
bmEvaluateHeaders/20       3040 ns         3040 ns       233345

this PR (if (strem_info == nullptr) {...} else {...}
bmEvaluateHeaders/2         159 ns          159 ns      4410206
bmEvaluateHeaders/4         690 ns          689 ns      1003884
bmEvaluateHeaders/6        1145 ns         1145 ns       698180
bmEvaluateHeaders/8        1354 ns         1354 ns       622681
bmEvaluateHeaders/10       1561 ns         1561 ns       486003
bmEvaluateHeaders/12       1762 ns         1762 ns       413769
bmEvaluateHeaders/14       2076 ns         2076 ns       338057
bmEvaluateHeaders/16       2405 ns         2405 ns       297029
bmEvaluateHeaders/18       2692 ns         2691 ns       262014
bmEvaluateHeaders/20       2990 ns         2990 ns       233128

this PR lambdas:
bmEvaluateHeaders/2         171 ns          171 ns      4083766
bmEvaluateHeaders/4         690 ns          690 ns      1002158
bmEvaluateHeaders/6        1154 ns         1154 ns       694188
bmEvaluateHeaders/8        1357 ns         1356 ns       606530
bmEvaluateHeaders/10       1571 ns         1571 ns       480932
bmEvaluateHeaders/12       1790 ns         1790 ns       403867
bmEvaluateHeaders/14       2085 ns         2085 ns       344280
bmEvaluateHeaders/16       2383 ns         2382 ns       297992
bmEvaluateHeaders/18       2679 ns         2679 ns       262552
bmEvaluateHeaders/20       3037 ns         3037 ns       236217
```

Reply:
It's fun optimizing stuff that matters :)

I will take a look later but I had a slightly different model in mind for how to use lambdas but let me look in detail at what you have when I get a bit of time.

Reply:
OK I just scanned over this again and I think this is what I would try, and why. I think the big-if-statement probably is what you want the compiler to generate for speed, but it's not readable/maintainable.

I think probably the lambdas are hard for the compiler to optimize away, and they are also not that easy to read -- they make this function very long.

I've thought of some other ways to do this but I am not confident the compiler will optimize them well. One I had been thinking was that the whole loop body in a lambda or function, which took the value as a formal. Then you'd have two copies of the for-loop but they would be one-liners calling the body one way or the other:

```
if (stream_info != nullptr) {
  for ([key, entry]. ,...) {
    handleKeyValue(request_headers, response_headers, stream_info, headers_to_overwrite, 
                   entry.formatter_->format(request_headers, response_headers, *stream_info));
  }
} else {
    for ([key, entry]. ,...) {
    handleKeyValue(request_headers, response_headers, &stream_info, &headers_to_overwrite, 
                   entry.original_value_);
  }
}
```
But I have another idea also. My main objection to your original loop was that you did a lot of pointless string copying in the case where stream_info==null. You could also avoid by using a string_view to hold the value, and a single string buffer declared outside the loop that you use to store the value when stream_info!=null.

Something like:
```
std::string value_buffer;
for (...) {
  absl::string_view value;
  if (stream_info != nullptr) {
    value_buffer = entry.formatter_->format(request_headers, response_headers, *stream_info);
    value = value_buffer;
  } else {
    value = entry.original_value_;
  }
  switch ...
}
```

I'll let you mull these options over; pick something that balances readability (ie functions not too long) and having decent performance.

One thing I'm not certain of: strings are actually pretty efficient these days if they are 22 bytes or less (search for "c++ small string optimization" -- I couldn't find any one authoritative post). So if the actual headers are mostly small your code is not that bad -- the main problem is the extra branch not the copies. But the moment you hit 23 bytes or 24 bytes  (I think) it needs a heap allocation. So depending on how you write your perf test you will get very different results.

Please pick something and document the tradeoffs in a comment and we can just go with it.

Reply:
Thanks for the great feedback. I think that your idea with using `absl::string_view`:
```
std::string value_buffer;
for (...) {
  absl::string_view value;
  if (stream_info != nullptr) {
    value_buffer = entry.formatter_->format(request_headers, response_headers, *stream_info);
    value = value_buffer;
  } else {
    value = entry.original_value_;
  }
  switch ...
}
```
strikes the best balance between readability and performance. 
I compared speed test results to `if {big block} else {big block}`  and new code results are almost identical for valid `*stream_info` and a bit worse for `stream_info` equal to `nullptr`.

As for small string optimizations, you are right that for clang, the cut off size is 22 bytes. The test results are different for long strings, but IMHO the performance hit  is unavoidable for any approach: _if {big block} else {big block}_, lambdas or the current code.